### PR TITLE
[go_router] Adds name to TypedGoRoute

### PR DESCRIPTION
**Pull request by dancamdev**
This PR is required for [#3665](https://github.com/flutter/packages/pull/3665) to work and be merged.

It should allow for generating `TypedGoRoutes` by defining a custom name and forwarding it to the underlying GoRoute. This is needed when working with Analytics services such as Google analytics that use the `RouteSettings.name` to log the page in their systems.

*List which issues are fixed by this PR. You must list at least one issue.*
[#120102](https://github.com/flutter/flutter/issues/120102)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by dancamdev** [comment]
@chunhtai is there anything blocking this?

File: packages/go_router/lib/src/route_data.dart [hunk 1]
```diff
@@ -233,6 +236,12 @@ class TypedGoRoute<T extends GoRouteData> extends TypedRoute<T> {
   ///
   final String path;
 
+  /// The name that corresponds to this route.
+  ///
+  /// See [GoRoute.name].
```
Comment (chunhtai) [inline packages/go_router/lib/src/route_data.dart:241]: Maybe mentions this is useful for analytics that reads the RouteSettings.name.

Comment (dancamdev) [reply packages/go_router/lib/src/route_data.dart:243]: Done.

**Review by hangyujin — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
can you fix the repo_checks? @dancamdev

**Comment by dancamdev** [comment]
@chunhtai should be good to go, when you get the chance! 👀

**Pull request merged**